packages/x-data-grid/src/hooks/features/pagination/useGridPagination.ts

```diff
@@ -29,7 +29,10 @@ export const paginationStateInitializer: GridStateInitializer<
 
   throwIfPageSizeExceedsTheLimit(paginationModel.pageSize, props.signature);
 
-  const rowCount = props.rowCount ?? props.initialState?.pagination?.rowCount;
+  const rowCount =
+    props.rowCount ??
+    props.initialState?.pagination?.rowCount ??
+    (props.paginationMode === 'client' ? state.rows?.totalRowCount : undefined);
```

Comment:
Why was this change necessary?

Reply:
Because otherwise the pagination test breaks. Pagination is dependent on rowCount event for client-side pagination, however it is never explicitly initialized.

The problem doesn't surface with the current SSR blocking, because the state is implicitly initiated in a useEffect somewhere, before the grid can mount. 

Or that's how I understood it anyways.

Reply:
Does that mean there is an occasion to remove a now redundant `useEffect`?

Reply:
Unfortunately not as a result of this change, this half could be changed if it could be tied to row updates for client-side pagination, but not sure if it's worth the effort. Since for server-side pagination, rowCount is not a derived state, but a standalone value that can be provided asynchronously, I'm not sure there's much to do about the useEffect:
https://github.com/mui/mui-x/blob/659dc5f42d3ba485c6dfdb7ae52f5e899bc66417/packages/x-data-grid/src/hooks/features/pagination/useGridRowCount.ts#L145-L147

Fwiw, this was the test that was breaking, was actually a cellSelection test, but pagination related still:
<img width="1488" alt="Screenshot 2024-12-02 at 20 33 32" src="https://github.com/user-attachments/assets/4c99de18-1fa9-4cd5-b124-1bff04b58e74">
<img width="1050" alt="Screenshot 2024-12-02 at 20 41 07" src="https://github.com/user-attachments/assets/64096103-75b3-4bba-b424-e6df7b5debf3">
<img width="1109" alt="Screenshot 2024-12-02 at 20 41 26" src="https://github.com/user-attachments/assets/09bf14a3-140d-4af8-9de6-5c9c6a0ebc71">